### DISCUSSION TITLE
TabbedContent (maybe other widgets) do not allow interaction unless composed first

### DISCUSSION THREAD

**Discussion by AAraKKe** [Textualize/textual#3266] [Q&A] [answered]
Hi, I have searched on opened issues about this but I cannot find anything so there are 2 options:

a) I am making a wrong assumption when using the library
b) There is an actually issue that should be fixed

If it is a) feel free to close 😅

## The issue
I was debugging my app and found that I cannot create a TabbedContent object and populate it as a I want and later yield it. This should be a quick oversimplified example that would break

```python
class TabbedApp(App[None]):
    def compose(self) -> ComposeResult:
        content = TabbedContent(id="content", initial="tabA")
        content.add_pane(TabPane("A", Label("This is tab A", id="tabA")))
        content.add_pane(TabPane("B", Label("This is tab B", id="tabB")))

        yield content
```

Th reason is that when we add a `TabPane` the method itself queries for the `Tabs` object which does not exisst in the Dom because we have never actually composed it. [This is the code](https://github.com/Textualize/textual/blob/85970972d9cb10afd00bb14f58335c9b8bdf99f9/src/textual/widgets/_tabbed_content.py#L305).

This creates, what seems to me, an unexpected side effect in which you cannot create your widgets freely and then render them because behaviour of the widget depends on whether they have been already rendered or not.

Maybe the intention of the `add_pane` method is to dynamically add panes after the `TabbedContent` has been generated but there doesn't seem to be another way of adding panes to the `TabbedContent` object since the constructor only supports titles.

I did what I usually do which is check the tests for the entity in case I can learn how to do it but the only test I find where add_pane is used is done after the compose has been run which would not highlight this issue. [This is the test](https://github.com/Textualize/textual/blob/85970972d9cb10afd00bb14f58335c9b8bdf99f9/tests/test_tabbed_content.py#L162).

Am I missing something?

#### Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.36.0 |
| Rich    | 13.5.2 |

#### Python

| Name           | Value                                                         |
|----------------|---------------------------------------------------------------|
| Version        | 3.11.4                                                        |
| Implementation | CPython                                                       |
| Compiler       | Clang 15.0.7                                                  |
| Executable     | /Users/jparaque/anaconda3/envs/learning-python/bin/python3.11 |

#### Operating System

| Name    | Value                                                                                            |
|---------|--------------------------------------------------------------------------------------------------|
| System  | Darwin                                                                                           |
| Release | 22.6.0                                                                                           |
| Version | Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64 |

#### Terminal

| Name                 | Value              |
|----------------------|--------------------|
| Terminal Application | iTerm.app (3.4.20) |
| TERM                 | xterm-256color     |
| COLORTERM            | truecolor          |
| FORCE_COLOR          | *Not set*          |
| NO_COLOR             | *Not set*          |

#### Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=254, height=33 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 254                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 33                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

Thanks!

**Comment by TomJGooding** [comment]
> Maybe the intention of the `add_pane` method is to dynamically add panes after the `TabbedContent` has been generated but there doesn't seem to be another way of adding panes to the `TabbedContent` object since the constructor only supports titles.

Do the examples in the Composing section of the guide to TabbedContents help? https://textual.textualize.io/widgets/tabbed_content/#composing

**Comment by willmcgugan** [comment, marked as answer]
You will need to compose the TabbedContent before you call `add_pane`.

Have you read the docs on [composing TabbedContent](https://textual.textualize.io/widgets/tabbed_content/#composing)?

If you need to use `add_pane` you could do that in `on_mount`. But you shouldn't need it unless its very dynamic.

**Comment by AAraKKe** [comment]
Thanks for the answers! Yes, I went't through the docs and saw the different ways composing it. What I think is misleading is the fact that it is not clear in code or clearly stated in the doc that certain actions require your widget to be composed, or at least I couldn't find it.

I think that what I found confusing is that in terms of OOP it seems possible to have a TabbedContent object to which you add a TabPane with the `add_pane` method. However, it is not because it needs to be composed first. Is there a section in the documentation where it is explained the composing process? Maybe that is what I missed and there it is explained that widgets generally require being composed first.
